In [1]:
## Q1. Explain the assumptions required to use ANOVA and provide examples of violations that could impact the validity of the results.
'''Assumptions of ANOVA:
(i) Normality of sampling distribution of mean:
The distribution of sample means is normally distributed.
(ii) Absence of outliers:
Outlying scores need to be removed from the dataset.
(iii) Homogeneity of variance:
Population variance in different levels of each independent variables should be equal.
(iv) Samples should be independent and randomly selected.
Potential assumption violations include:
Implicit factors: lack of independence within a sample
Outliers: apparent nonnormality by a few data points
Nonnormality: nonnormality of entire samples
Unequal population variances
Patterns in plots of data: detecting violation assumptions graphically
Special problems with small sample sizes
Special problems with unbalanced sample sizes
Multiple comparisons: effects of assumption violations on multiple comparison tests'''

'Assumptions of ANOVA:\n(i) Normality of sampling distribution of mean:\nThe distribution of sample means is normally distributed.\n(ii) Absence of outliers:\nOutlying scores need to be removed from the dataset.\n(iii) Homogeneity of variance:\nPopulation variance in different levels of each independent variables should be equal.\n(iv) Samples should be independent and randomly selected.\nPotential assumption violations include:\nImplicit factors: lack of independence within a sample\nOutliers: apparent nonnormality by a few data points\nNonnormality: nonnormality of entire samples\nUnequal population variances\nPatterns in plots of data: detecting violation assumptions graphically\nSpecial problems with small sample sizes\nSpecial problems with unbalanced sample sizes\nMultiple comparisons: effects of assumption violations on multiple comparison tests'

In [2]:
## Q2. What are the three types of ANOVA, and in what situations would each be used?
'''(i) One-way ANOVA:
It has one factor with at least two levels and these levels are independent.
e.g.: Doctor wants to test a new medication to decrease the headache. He splits the participants in 3 conditions [10g,20mg,30mg] and ask them to rate the headache between 1 to 10.
Here,
Medication ---->Factor
[10g,20mg,30mg] ---->Independent levels

(ii) Repeated measures ANOVA:
It has one factor with at least two levels where levels are dependent.
e.g.: A person running on different days (day1, day2, day3) are measured.
Here,
Running ----> Factor
(day1, day2, day3) ----> Dependent levels since same person runs on different days.

(iii) Factorial ANOVA:
It has two or more factors, each of which with at least two levels and this level can be independent and dependent.
e.g.: Persons (Male, Female) running on different days (day1, day2, day3) are measured.
Here,
Running ----> Factor
(day1, day2, day3) ----> Dependent levels since same person runs on different days.
(Male, Female) ---->Independent levels.'''

'(i) One-way ANOVA:\nIt has one factor with at least two levels and these levels are independent.\ne.g.: Doctor wants to test a new medication to decrease the headache. He splits the participants in 3 conditions [10g,20mg,30mg] and ask them to rate the headache between 1 to 10.\nHere,\nMedication ---->Factor\n[10g,20mg,30mg] ---->Independent levels\n\n(ii) Repeated measures ANOVA:\nIt has one factor with at least two levels where levels are dependent.\ne.g.: A person running on different days (day1, day2, day3) are measured.\nHere,\nRunning ----> Factor\n(day1, day2, day3) ----> Dependent levels since same person runs on different days.\n\n(iii) Factorial ANOVA:\nIt has two or more factors, each of which with at least two levels and this level can be independent and dependent.\ne.g.: Persons (Male, Female) running on different days (day1, day2, day3) are measured.\nHere,\nRunning ----> Factor\n(day1, day2, day3) ----> Dependent levels since same person runs on different days.\n(Male, F

In [3]:
## Q3. What is the partitioning of variance in ANOVA, and why is it important to understand this concept?
''' Partitioning of variance in ANOVA is the hypothesis testing in ANOVA.
A more important use of the F-distribution is in analyzing variance to see if three or more samples come from populations with equal means. 
This is an important statistical test, not so much because it is frequently used, 
but because it is a bridge between univariate statistics and multivariate statistics and because the strategy it uses is one which is used in many multivariate tests and procedures.'''

' Partitioning of variance in ANOVA is the hypothesis testing in ANOVA.\nA more important use of the F-distribution is in analyzing variance to see if three or more samples come from populations with equal means. \nThis is an important statistical test, not so much because it is frequently used, \nbut because it is a bridge between univariate statistics and multivariate statistics and because the strategy it uses is one which is used in many multivariate tests and procedures.'

In [4]:
## Q4. How would you calculate the total sum of squares (SST), explained sum of squares (SSE), and residual sum of squares (SSR) in a one-way ANOVA using Python?
'''1. Sum of Squares Total (SST) – The sum of squared differences between individual data points (yi) and the mean of the response variable (y).
SST = Σ(yi – y)2
2. Sum of Squares Regression (SSR) – The sum of squared differences between predicted data points (ŷi) and the mean of the response variable(y).
SSR = Σ(ŷi – y)2
3. Sum of Squares Error (SSE) – The sum of squared differences between predicted data points (ŷi) and observed data points (yi).
SSE = Σ(ŷi – yi)2
The following relationship exists between these three measures:
SST = SSR + SSE'''

'1. Sum of Squares Total (SST) – The sum of squared differences between individual data points (yi) and the mean of the response variable (y).\nSST = Σ(yi – y)2\n2. Sum of Squares Regression (SSR) – The sum of squared differences between predicted data points (ŷi) and the mean of the response variable(y).\nSSR = Σ(ŷi – y)2\n3. Sum of Squares Error (SSE) – The sum of squared differences between predicted data points (ŷi) and observed data points (yi).\nSSE = Σ(ŷi – yi)2\nThe following relationship exists between these three measures:\nSST = SSR + SSE'

In [5]:
## Q5. In a two-way ANOVA, how would you calculate the main effects and interaction effects using Python?
'''Main effects deal with each factor separately.The interaction is the simultaneous changes in the levels of both factors.'''
import statsmodels.api as sm
from statsmodels.formula.api import ols
import pandas as pd
yellow_placebo = [2, 4, 3, 1, 5 ]
yellow_fuel = [4, 7, 6, 6, 6]
blue_placebo = [5, 6, 4, 2, 8 ]
blue_fuel = [5, 6, 4, 7, 8]
red_placebo = [1, 7, 5, 2, 4]
red_fuel = [11, 13, 12, 14, 10]
data = []
for i in yellow_placebo:
    data.append((i,"yellow","N"))
for i in yellow_fuel:
    data.append((i, "yellow", "Y"))
for i in blue_placebo:
    data.append((i, "blue", "N"))
for i in blue_fuel:
    data.append((i, "blue","Y"))
for i in red_placebo:
    data.append((i, "red","N"))
for i in red_fuel:
    data.append((i, "red","Y"))   
columns = ['Percent_Muscle_Increase','Shell_Color','Snail_Fuel']
snail_fuel_data = pd.DataFrame(data,columns=columns)
#perform two-way ANOVA
model = ols('Percent_Muscle_Increase ~ C(Snail_Fuel) + C(Shell_Color) + C(Snail_Fuel):C(Shell_Color)', data=snail_fuel_data).fit()
sm.stats.anova_lm(model, typ=3)

sum_sq    df          F    PR(>F)
Intercept                     125.000000   1.0  38.659794  0.000002
C(Snail_Fuel)                   2.500000   1.0   0.773196  0.387950
C(Shell_Color)                 10.133333   2.0   1.567010  0.229279
C(Snail_Fuel):C(Shell_Color)   70.200000   2.0  10.855670  0.000439
Residual                       77.600000  24.0        NaN       NaN

#Observation:
- The model tells us that the only significant effect in our ANOVA is the interaction of shell color and snail fuel. The p values (PR(>F)) for the snail fuel and shell color variables on their own would not meet the <.05 standard of significance.
- Given these results, we can say that for the total population of snails in the study, there was no statistically significant difference in muscle increase between those who drank Snail Fuel and those who drank a placebo energy drink. However, there was a statistically significant interaction between shell color and drinking Snail Fuel, such that snails with red shells saw larger gains in muscle mass than snails with yellow and blue shells.

In [6]:
## Q6. Suppose you conducted a one-way ANOVA and obtained an F-statistic of 5.23 and a p-value of 0.02. What can you conclude about the differences between the groups, and how would you interpret these results?
'''Since F-statistic of 5.23 is greater than p-value of 0.02, we reject the null hypothesis that states the difference of mean between the two group is same.'''

'Since F-statistic of 5.23 is greater than p-value of 0.02, we reject the null hypothesis that states the difference of mean between the two group is same.'

In [7]:
## Q7. In a repeated measures ANOVA, how would you handle missing data, and what are the potential consequences of using different methods to handle missing data?
'''Multiple imputation:
One of the most effective ways of dealing with missing data is multiple imputation (MI). 
Using MI, we can create multiple plausible replacements of the missing data, given what we have observed and a statistical model (the imputation model).
In the ANOVA, using MI has the additional benefit that it allows taking covariates into account that are relevant for the missing data but not for the analysis. 
Listwise deletion:
The result is a direct consequence of how the missing data were simulated. Fortunately, there are statistical methods that can account for the missing data and help us obtain more trustworthy results.'''

'Multiple imputation:\nOne of the most effective ways of dealing with missing data is multiple imputation (MI). \nUsing MI, we can create multiple plausible replacements of the missing data, given what we have observed and a statistical model (the imputation model).\nIn the ANOVA, using MI has the additional benefit that it allows taking covariates into account that are relevant for the missing data but not for the analysis. \nListwise deletion:\nThe result is a direct consequence of how the missing data were simulated. Fortunately, there are statistical methods that can account for the missing data and help us obtain more trustworthy results.'

In [8]:
## Q8. What are some common post-hoc tests used after ANOVA, and when would you use each one? Provide an example of a situation where a post-hoc test might be necessary.
'''In order to find out exactly which groups are different from each other, 
we must conduct a post hoc test (also known as a multiple comparison test), which will allow us to explore the difference between multiple group means while also controlling for the family-wise error rate.
It’s important to note that we only need to conduct a post hoc test when the p-value for the ANOVA is statistically significant. If the p-value is not statistically significant, 
this indicates that the means for all of the groups are not different from each other, so there is no need to conduct a post hoc test to find out which groups are different from each other.
Post-hoc tests used after ANOVA:
(i) Tukey’s Test – useful when you want to make every possible pairwise comparison.We can perform Tukey’s Test for multiple comparisons by using the built-in R function TukeyHSD()
(ii) Holm’s Method – a slightly more conservative test compared to Tukey’s Test
(iii) Dunnett’s Correction – useful when you want to compare every group mean to a control mean, and you’re not interested in comparing the treatment means with one another.
Post hoc tests do a great job of controlling the family-wise error rate, but the tradeoff is that they reduce the statistical power of the comparisons. 
This is because the only way to lower the family-wise error rate is to use a lower significance level for all of the individual comparisons.
The problem with this is that lower significance levels correspond to lower statistical power. 
This means that if a difference between group means actually does exist in the population, a study with lower power is less likely to detect it. '''

'In order to find out exactly which groups are different from each other, \nwe must conduct a post hoc test (also known as a multiple comparison test), which will allow us to explore the difference between multiple group means while also controlling for the family-wise error rate.\nIt’s important to note that we only need to conduct a post hoc test when the p-value for the ANOVA is statistically significant. If the p-value is not statistically significant, \nthis indicates that the means for all of the groups are not different from each other, so there is no need to conduct a post hoc test to find out which groups are different from each other.\nPost-hoc tests used after ANOVA:\n(i) Tukey’s Test – useful when you want to make every possible pairwise comparison.We can perform Tukey’s Test for multiple comparisons by using the built-in R function TukeyHSD()\n(ii) Holm’s Method – a slightly more conservative test compared to Tukey’s Test\n(iii) Dunnett’s Correction – useful when you want 

In [9]:
'''Q9. A researcher wants to compare the mean weight loss of three diets: A, B, and C. They collect data from 50 participants who were randomly assigned to one of the diets. Conduct a one-way ANOVA using Python
to determine if there are any significant differences between the mean weight loss of the three diets. Report the F-statistic and p-value, and interpret the results.'''
import pandas
import random
from scipy.stats import f_oneway
random.seed(10)
d1=[random.randint(10,32) for i in range(50)]
d2=[random.randint(10,32) for i in range(50)]
d3=[random.randint(10,32) for i in range(50)]
f_oneway(d1,d2,d3)

F_onewayResult(statistic=0.6579128003050717, pvalue=0.5194494896494897)

#Observation:
Since F statistic=0.6579128003050717is greater than pvalue=0.5194494896494897), we reject the null hypothesis that states all the mean weight loss of three diets are equal.


In [10]:
'''Q10. A company wants to know if there are any significant differences in the average time it takes to complete a task using three different software programs: Program A, Program B, and Program C. They
randomly assign 30 employees to one of the programs and record the time it takes each employee to complete the task. Conduct a two-way ANOVA using Python to determine if there are any main effects or
interaction effects between the software programs and employee experience level (novice vs experienced). Report the F-statistics and p-values, and interpret the results.'''
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
# Load data
data = pd.read_csv('Software_Data.csv')

# Fit the model
model = ols('time ~ C(program) + C(experience) + C(program):C(experience)', data=data).fit()
# Perform two-way ANOVA
aov_table = sm.stats.anova_lm(model, typ=2)
# Print the results
print(aov_table)

                            sum_sq    df          F        PR(>F)
C(program)                4.394667   2.0  30.948357  2.263687e-07
C(experience)             0.065333   1.0   0.920188  3.469906e-01
C(program):C(experience)  0.082667   2.0   0.582160  5.663862e-01
Residual                  1.704000  24.0        NaN           NaN


In [11]:
'''Q11. An educational researcher is interested in whether a new teaching method improves student test scores. They randomly assign 100 students to either the control group (traditional teaching method) or the
experimental group (new teaching method) and administer a test at the end of the semester. Conduct a two-sample t-test using Python to determine if there are any significant differences in test scores
between the two groups. If the results are significant, follow up with a post-hoc test to determine whichgroup(s) differ significantly from each other.'''
import numpy as np
from scipy.stats import ttest_ind
# Create two random samples
sample1 = np.random.normal(0, 1, 100)
sample2 = np.random.normal(1, 1, 100)
# Conduct a two-sample t-test
t_stat, p_val = ttest_ind(sample1, sample2)
print('t-statistic:', t_stat)
print('p-value:', p_val)

t-statistic: -7.603743227746189
p-value: 1.131642205443929e-12


In [12]:
'''Q12. A researcher wants to know if there are any significant differences in the average daily sales of three retail stores: Store A, Store B, and Store C. They randomly select 30 days and record the sales for each store
on those days. Conduct a repeated measures ANOVA using Python to determine if there are any significant differences in sales between the three stores. If the results are significant, follow up with a post-
hoc test to determine which store(s) differ significantly from each other.'''
import pandas as pd
from statsmodels.stats.anova import AnovaRM
data = pd.read_csv('Store_sales.csv')
anovarm = AnovaRM(data,'sales','store', within=['day'])
fit = anovarm.fit()
print(fit.summary())

              Anova
    F Value  Num DF  Den DF Pr > F
----------------------------------
day  0.8446 29.0000 58.0000 0.6847

